storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java

```diff
@@ -178,53 +181,74 @@
-        QueryRequest queryRequest) {
+        QueryRequest request) {
       this.tracesStore = tracesStore;
       this.traceIdsByTsStore = traceIdsByTsStore;
-      this.queryRequest = queryRequest;
+      this.request = request;
     }
 
     @Override public List<List<Span>> query() {
       List<List<Span>> traces = new ArrayList<>();
       List<String> traceIds = new ArrayList<>();
       // milliseconds to microseconds
-      long from = (queryRequest.endTs() - queryRequest.lookback()) * 1000;
-      long to = queryRequest.endTs() * 1000;
-      // first index
-      KeyValueIterator<Long, Set<String>> spanIds = traceIdsByTsStore.range(from, to);
-      spanIds.forEachRemaining(keyValue -> {
-        for (String traceId : keyValue.value) {
-          if (!traceIds.contains(traceId) && traces.size() < queryRequest.limit()) {
-            List<Span> spans = tracesStore.get(traceId);
-            if (spans != null && queryRequest.test(spans)) { // apply filters
-              traceIds.add(traceId); // adding to check if we have already add it later
-              traces.add(spans);
+      long from = (request.endTs() - request.lookback()) * 1000;
+      long to = request.endTs() * 1000;
+      long checkpoint = to - (30 * 1000 * 1000); // 30 sec before upper bound
+      if (checkpoint <= from) { // do one run
+        try (KeyValueIterator<Long, Set<String>> spanIds = traceIdsByTsStore.range(from, to)) {
+          spanIds.forEachRemaining(keyValue -> {
+            for (String traceId : keyValue.value) {
+              if (!traceIds.contains(traceId)) {
+                List<Span> spans = tracesStore.get(traceId);
+                if (spans != null && request.test(spans)) { // apply filters
+                  traceIds.add(traceId); // adding to check if we have already add it later
+                  traces.add(spans);
+                }
+              }
             }
+          });
+        }
+      } else {
+        while (checkpoint > from && traces.size() < request.limit()) {
+          try (KeyValueIterator<Long, Set<String>> spanIds = traceIdsByTsStore.range(checkpoint, to)) {
+            spanIds.forEachRemaining(keyValue -> {
+              for (String traceId : keyValue.value) {
+                if (!traceIds.contains(traceId)) {
+                  List<Span> spans = tracesStore.get(traceId);
+                  if (spans != null && request.test(spans)) { // apply filters
+                    traceIds.add(traceId); // adding to check if we have already add it later
+                    traces.add(spans);
+                  }
+                }
+              }
+            });
           }
+          to = checkpoint;
+          checkpoint = checkpoint - (60 * 1000 * 1000); // 1 min before more
         }
-      });
-      LOG.debug("Traces found from query {}: {}", queryRequest, traces.size());
-      return traces;
+      }
+      traces.sort(Comparator.<List<Span>>comparingLong(o -> o.get(0).timestampAsLong()).reversed());
```

Comment:
Resulting traces list should have values close to `limit`, so sorting should not be expensive.